internal/verifier/change_stream.go

```diff
@@ -67,121 +82,233 @@ func (verifier *Verifier) GetChangeStreamFilter() []bson.D {
+			return nil
+		}
+
+		err := verifier.persistChangeStreamResumeToken(ctx, cs)
+		if err == nil {
+			lastPersistedTime = time.Now()
+		}
+
+		return err
+	}
+
+	for {
+		select {
+		// if the context is cancelled return immmediately
+		case <-ctx.Done():
+			return
+		// if the changeStreamEnderChan has a message, we have moved to the Recheck phase, obtain
+		// the remaining changes, but when TryNext returns false, we will exit, since there should
+		// be no message until the user has guaranteed writes to the source have ended.
+		case <-verifier.changeStreamEnderChan:
+			for cs.TryNext(ctx) {
 				if err := cs.Decode(&changeEvent); err != nil {
-					verifier.logger.Fatal().Err(err).Msg("")
+					verifier.logger.Fatal().Err(err).Msg("Failed to decode change event")
 				}
 				err := verifier.HandleChangeStreamEvent(ctx, &changeEvent)
 				if err != nil {
 					verifier.changeStreamErrChan <- err
-					return
+					verifier.logger.Fatal().Err(err).Msg("Error handling change event")
+				}
+			}
+			verifier.mux.Lock()
+			verifier.changeStreamRunning = false
+			if verifier.lastChangeEventTime != nil {
+				verifier.srcStartAtTs = verifier.lastChangeEventTime
+			}
+			verifier.mux.Unlock()
+			// since we have started Recheck, we must signal that we have
+			// finished the change stream changes so that Recheck can continue.
+			verifier.changeStreamDoneChan <- struct{}{}
+			// since the changeStream is exhausted, we now return
+			verifier.logger.Debug().Msg("Change stream is done")
+			return
+		// the default case is that we are still in the Check phase, in the check phase we still
+		// use TryNext, but we do not exit if TryNext returns false.
+		default:
+			var err error
+
+			if next := cs.TryNext(ctx); next {
+				if err = cs.Decode(&changeEvent); err != nil {
+					err = errors.Wrapf(err, "failed to decode change event (%v)", cs.Current)
+				}
+
+				if err == nil {
+					err = verifier.HandleChangeStreamEvent(ctx, &changeEvent)
+					if err != nil {
+						err = errors.Wrapf(err, "failed to handle change event (%+v)", changeEvent)
+					}
 				}
 			}
+
+			if err == nil {
+				err = persistResumeTokenIfNeeded()
+			}
+
+			if err != nil {
+				verifier.changeStreamErrChan <- err
+				return
+			}
 		}
 	}
+}
+
+// StartChangeStream starts the change stream.
+func (verifier *Verifier) StartChangeStream(ctx context.Context) error {
 	pipeline := verifier.GetChangeStreamFilter()
 	opts := options.ChangeStream().SetMaxAwaitTime(1 * time.Second)
-	if startTime != nil {
-		opts = opts.SetStartAtOperationTime(startTime)
-		verifier.srcStartAtTs = startTime
+
+	savedResumeToken, err := verifier.loadChangeStreamResumeToken(ctx)
+	if err != nil {
+		return errors.Wrap(err, "failed to load persisted change stream resume token")
 	}
+
+	csStartLogEvent := verifier.logger.Info()
+
+	if savedResumeToken != nil {
+		logEvent := csStartLogEvent.
+			Stringer("resumeToken", savedResumeToken)
```

Comment:
I would think that not to happen here because the verifier will run simpler change streams than mongosync does under ORR, but yeah it’s probably best not to chance it.

Replacing with an extracted timestamp.

Reply:
Even so, it seems best to err here on the side of caution. We probably don’t need the full resume token in the log.